### PR TITLE
Update: add source property to LintResult object (fixes #7098)

### LINKED ISSUE #7098: Proposal: add `sourceContext` property to the messages generated by the `eslint` object
**Context:**
We would like to port the JSCS default formatter over to ESLint. That formatter prints out the lines around the location of an error/warning. See #5860 for an example output.

**Problem**
Unfortunately, with ESLint, there is currently no way to get those lines of "context" from a formatter, because:

1. The formatter is invoked with only one argument - the array of messages to report on, and those do not contain any information about the lines of context around the error.
2. The formatter is intentionally made to be "dumb" and shouldn't/cannot use `SourceCode.getLines()` to retrieve the lines of code of the file being reported on.

**Proposal**
I would like to add a `sourceContext` property to each message object that gets returned by the `eslint.verify()` method.
With this new property, the message object would look roughly like the following:

```javascript
{
    ruleId: 'lines-around-comment',
    severity: 2,
    message: 'Expected line before comment.',
    line: 37,
    column: 9,
    nodeType: 'Line',
    source: '// some random comment',
    sourceContext: {
        before: [
            "line of code before line of the error",
            "another line of code"
        ],
        after: [
            "line of code after the line of the error",
            "another line of code"
        ]
    }
}
```

Each item of the `before`/`after` array is a line of code directly from the array returned by `SourceCode.getLines(code)`.

An alternative would be that each item is instead an object that contains a `source` and a `line` property, like this:

```javascript
{
    sourceContext: {
        before: [
            {
                source: "line of code before line of the error",
                line: 35
            },
            {
                source: "another line of code",
                line: 36
            }
        ],
        after: [
            // ...
        ]
    }
}
```
The advantage of this alternative approach is that it would prevent delegating the calculation of the line number to the formatter.

Adding this `sourceContext` property would solve the problem described, and would allow us to port the JSCS formatter with ease.
I have some proof-of-concept code for this proposal locally that is already able to print out the lines around an error, like this:

```bash
$ bin/eslint.js --rulesdir lib/internal-rules/ lib/formatters/stylish.js
35
36     results.forEach(function(result) {
37         // console.log(result);
38         const messages = result.messages;
39

lib/formatters/stylish.js
  37:9  error  Expected line before comment  lines-around-comment
```

The existing tests also already pass so I would only need to clean up the code a bit and make sure to test some edge cases better.

**Questions**
- Are there other any other ways to solve the problem described that would be preferable?
- I don't really like the property name `sourceContext` that much, but couldn't come up with anything better. Any suggestions?

### PR DESCRIPTION
**Pull request by vitorbal**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

This PR refers to #7098.

**Please check each item to ensure your pull request is ready:**

- [x] I've read the [pull request guide](http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [x] I've included tests for my change
- [x] I've updated documentation for my change (if appropriate)

**What changes did you make? (Give an overview)**
Add a source property containing the entire file contents as a string into the linting results for each file that has linting errors. This can be used by formatters to output source code along with lint messages. This is a non-breaking change.

**Is there anything you'd like reviewers to focus on?**
A couple of important API decisions were done during the development of this PR in an effort to minimize redundancy and memory usage, and I would like to know if the rest of the team agrees with them:

- The `source` property is omitted if no errors or warnings are present. As there is nothing to report, there is no need for the source code to be present.
- The `source` property is also omitted if there were fixes and a `output` property is present. In this case, the `source` property would be redundant as the `output` property already contains the source code (also, the messages' `column` and `line` properties are updated to correspond to the location on the source code after the fixes)

### CODE REVIEW

File: docs/developer-guide/nodejs-api.md [hunk 1]
```diff
@@ -240,7 +240,15 @@ The return value is an object containing the results of the linting operation. H
 }
 ```
 
-The top-level report object has a `results` array containing all linting results for files that had warnings or errors (any files that did not produce a warning or error are omitted). Each file result includes the `filePath`, a `messages` array, `errorCount`, `warningCount`, and optionally `output`. The `messages` array contains the result of calling `linter.verify()` on the given file. The `errorCount` and `warningCount` give the exact number of errors and warnings respectively on the given file. The `output` property gives the source code for the file with as many fixes applied as possible, so you can use that to rewrite the files if necessary. The top-level report object also has `errorCount` and `warningCount` which give the exact number of errors and warnings respectively on all the files.
+The top-level report object has a `results` array containing all linting results for files that had warnings or errors (any files that did not produce a warning or error are omitted). Each file result includes:
```
Comment (vitorbal) [inline docs/developer-guide/nodejs-api.md]: Did some restructuring here while adding the documentation for the new `source` property, hope it's okay.

File: docs/developer-guide/nodejs-api.md [hunk 2]
```diff
@@ -240,7 +240,15 @@ The return value is an object containing the results of the linting operation. H
...
+
+* `filePath` - Path to the given file.
+* `messages` - Array containing the result of calling `linter.verify()` on the given file.
+* `errorCount` and `warningCount` - The exact number of errors and warnings respectively on the given file.
+* `source` - The source code for the given file. This property is omitted if this file has no errors/warnings or if the `output` property is present.
```
Comment (nzakas) [inline docs/developer-guide/nodejs-api.md]: Hmmm, I'm not sure if omitting this when there's output is expected. Can you explain your thinking a bit more? I'm just worried about formatters all needing to check for both.

Comment (vitorbal) [inline docs/developer-guide/nodejs-api.md]: Thanks for the question!
My reasoning was that when fixes are applied, the `source` property would be redundant, since the `output` property already contains the source code.
Essentially, I thought we could omit the `source` property in that case in an effort to reduce memory usage. I do agree it complicates the API though, so I would be fine with not omitting  in that case as well if we prefer.

Comment (nzakas) [inline docs/developer-guide/nodejs-api.md]: Wouldn't source be the text before fixes are applied?

Comment (vitorbal) [inline docs/developer-guide/nodejs-api.md]: Sorry, I should have explained this as well from the get-go. Apologies for not being clear enough.

In my initial implementation, `source` was indeed the text before fixes.
But all of the error/warning messages in the `messages` property contain the line/column location **after** the fixes, so if we would use the `source` property in that case, the formatter would potentially report an incorrect location for each error.

As an example, let's say there are two messages, the first can be fixed, the second one cannot:
```javascript
[{
    message: 'Expected line before comment',
    line: 4,
    column: 1
},
{
    message: '\'foo\' is defined but never used.',
    line: 5,
    column: 7
}]
```

After the first message is fixed, a newline is added above line 4 in the source code. The messages array is then updated to contain only one message:

```javascript
[{
    message: '\'foo\' is defined but never used.',
    line: 6, // <--- line is updated to reflect fixes
    column: 7
}]
```

-- 

This means that the formatter would have to check for the presence of `output` in that case anyway, even if the `source` property would not be omitted when there are fixes.

Comment (nzakas) [inline docs/developer-guide/nodejs-api.md]: Ah, I see, thanks for explaining. Okay, you've convinced me that it's okay to leave off `source` because of this. Can you please update the custom formatter documentation to make note of this?

**Comment by nzakas** [comment]
@vitorbal I think this is probably ready after the docs changes, but I noticed you add the "do not merge" label. If you feel like you're ready, please remove that label.

**Comment by vitorbal** [comment]
@nzakas just pushed a new version that updates the custom formatters documentation!

I only added the "do not merge" label since the original issue (#7098) is still not marked as accepted. Do you think we could mark it as accepted now? Seems like no one manifested anything against the proposal after you posted the design summary.

**Comment by ilyavolodin** [comment]
I marked original issue as `accepted` as there's enough support on that thread from the members of the team.

**Comment by vitorbal** [comment]
@ilyavolodin awesome, thanks!

**Comment by vitorbal** [comment]
Pushed another version that fine-tunes the custom formatters doc changes a bit.

File: docs/developer-guide/working-with-custom-formatters.md [hunk 3]
```diff
@@ -67,9 +68,11 @@ the list of messages for `errors` and/or `warnings`.
...
 
 - **filePath**: The path to the file relative to the current working directory (the path from which eslint was executed).
-- **messages**: An array of message objects. See below for more info about messages
-- **errorCount**: The number of errors for the given file
-- **warningCount**: the number of warnings for the give file
+- **messages**: An array of message objects. See below for more info about messages.
+- **errorCount**: The number of errors for the given file.
+- **warningCount**: The number of warnings for the given file.
+- **source**: The source code for the given file. This property is omitted if this file has no errors/warnings or if the `output` property is present.
+- **output**: The source code for the given file with as many fixes applied as possible. This property is omitted if no fix is available.
```
Comment (nzakas) [inline docs/developer-guide/working-with-custom-formatters.md]: I'm not sure where, but we should probably mention that the `source` property on the individual messages will be going away and that people should instead use `source` or `output` from the top-level of the structure.

**Comment by nzakas** [comment]
@vitorbal looks good!

@btmills @mysticatea I think we should call this out in the release notes.

**Comment by vitorbal** [comment]
@nzakas thanks for all the feedback on this one!

File: docs/developer-guide/nodejs-api.md (hunk 2 above)
Comment (SimenB) [inline docs/developer-guide/nodejs-api.md]: Why does it specify that it's  "omitted if this file has no errors/warnings", when the whole node is gone if no err/warns as noted in the paragraph above ("any files that did not produce a warning or error are omitted")?
